hack/e2e/cmd/generateServicePrincipals.go

```diff
@@ -0,0 +1,218 @@
+	"github.com/Azure/ARO-RP/pkg/util/keyvault"
+	utillog "github.com/Azure/ARO-RP/pkg/util/log"
+)
+
+// generateServicePrincipalsCmd represents the generateServicePrincipals command
+var generateServicePrincipalsCmd = &cobra.Command{
+	Use:   "generateServicePrincipals",
+	Short: "Generate the service principals for E2E",
+	RunE: func(cmd *cobra.Command, args []string) error {
+		err := generateKeys(cmd)
+		if err != nil {
+			cmd.PrintErr(err)
+			return err
+		}
+		return nil
+	},
+}
+
+func init() {
+	rootCmd.AddCommand(generateServicePrincipalsCmd)
+
+	generateServicePrincipalsCmd.Flags().String("keyvault", "aro-e2e-principals", "keyvault to use")
+	generateServicePrincipalsCmd.Flags().String("spn-prefix", "aro-v4-e2e-devops-spn-", "prefix for SPN name")
+	generateServicePrincipalsCmd.Flags().Int("spn-count", 50, "number of SPNs to generate/use")
+}
+
+func generateKeys(cmd *cobra.Command) error {
+	ctx := cmd.Context()
+	log := utillog.GetLogger()
+
+	_env, err := env.NewCore(ctx, log, env.COMPONENT_TOOLING)
+	if err != nil {
+		return err
+	}
+
+	keyvaultParam, err := cmd.Flags().GetString("keyvault")
+	if err != nil {
+		return err
+	}
+
+	spnPrefixParam, err := cmd.Flags().GetString("spn-prefix")
+	if err != nil {
+		return err
+	}
+
+	spnCountParam, err := cmd.Flags().GetInt("spn-count")
+	if err != nil {
+		return err
+	}
+
+	azcred, err := azidentity.NewDefaultAzureCredential(nil)
+	if err != nil {
+		return err
+	}
+
+	cred := azidext.NewTokenCredentialAdapter(azcred, []string{_env.Environment().KeyVaultScope})
+
+	keyvaultURI := keyvault.URI(_env, keyvaultParam, "")
+	keyvaultManager := keyvault.NewManager(cred, keyvaultURI)
+
+	graphClient, err := _env.Environment().NewGraphServiceClient(azcred)
+	if err != nil {
+		return err
+	}
+
+	for i := 1; i == spnCountParam; i++ {
+		spnObjectID := ""
+		spnAppID := ""
+		spnName := fmt.Sprintf("%s%d", spnPrefixParam, i)
+
+		log.Infof("Processing spn %s\n", spnName)
+
+		log.Info("Looking up object ID...")
+
+		filter := fmt.Sprintf("displayName eq '%s'", spnName)
+		result, err := graphClient.Applications().Get(ctx, &msgraph_apps.ApplicationsRequestBuilderGetRequestConfiguration{
+			QueryParameters: &msgraph_apps.ApplicationsRequestBuilderGetQueryParameters{
+				Filter: &filter,
+				Select: []string{"id", "appId", "passwordCredentials"},
+			},
+		})
+		if err != nil {
+			return err
+		}
+
+		apps := result.GetValue()
+		switch len(apps) {
+		case 0:
+			return fmt.Errorf("no applications found for spn '%s'", spnName)
+		case 1:
+			spnObjectID = *apps[0].GetId()
+			spnAppID = *apps[0].GetAppId()
+		default:
+			return fmt.Errorf("%d applications found for spn '%s'", len(apps), spnName)
+		}
+
+		log.Infof("found IDs for '%s': ObjectID='%s', AppID='%s'", spnName, spnObjectID, spnAppID)
+
+		log.Info("Ensuring SPN ID in keyvault...")
+
+		err = keyvaultManager.SetSecret(ctx, fmt.Sprintf("%s-app-id", spnName), azkeyvault.SecretSetParameters{
+			Value: to.StringPtr(spnAppID),
+		})
```

Comment:
IMHO extracting into a variable inline structs that are > 1 line long is a good thing (talking about azkeyvault.SecretSetParameters{)